packages/compiler-cli/src/ngtsc/metadata/src/field_mapping.ts

```diff
@@ -0,0 +1,185 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {InputOutputPropertySet} from '@angular/compiler';
+
+type FieldName = string;
+type PropertyName = string;
+
+/**
+ * An input or output of a directive that has both a JavaScript field name as well as an Angular
+ * template property name.
+ */
+export interface InputOrOutput {
+  /**
+   * The field name on the component or directive instance for this input or output.
+   */
+  fieldName: string;
+
+  /**
+   * The property name used to bind this input or output in an Angular template.
+   */
+  propertyName: string;
+}
+
+/**
+ * A mapping of field and property names, such as that for a particular directive or component.
+ *
+ * A single field name has exactly one input/output (and therefore one property name) associated
+ * with it, but the same property name may be shared across many field names.
+ *
+ * Allows bidirectional querying of the mapping - looking up all inputs/outputs with a given
+ * property name, or mapping from a specific field to its property name.
+ */
+export class FieldPropertyMapping implements InputOutputPropertySet {
+  /**
+   * Mapping from field names to the single `InputOrOutput` for that field.
+   */
+  private forwardMap: Map<FieldName, InputOrOutput>;
+
+  /**
+   * Mapping from property names to one or more `InputOrOutput`s which share that name.
+   */
+  private reverseMap: Map<PropertyName, InputOrOutput[]>;
+
+  private constructor(forwardMap: Map<FieldName, InputOrOutput>) {
+    this.forwardMap = forwardMap;
+    this.reverseMap = reverseMapFromForwardMap(forwardMap);
+  }
+
+  /**
+   * Construct a `FieldPropertyMapping` with no entries.
+   */
+  static empty(): FieldPropertyMapping {
+    return new FieldPropertyMapping(new Map());
+  }
+
+  /**
+   * Construct a `FieldPropertyMapping` from a primitive JS object which maps field names to either
+   * property names or an array that contains both names, which is used in on-disk metadata formats
+   * (e.g. in .d.ts files).
+   */
+  static fromFieldMappedObject(
+      obj: {[fieldName: string]: string|[/* propertyName */ string, /* fieldName */ string]}):
```

Comment:
```suggestion
      obj: {[fieldName: string]: PropertyName|[PropertyName, FieldName]}):
```
Should this be something like that?

Reply:
This is a public interface, so the alias types aren't used.